core/src/main/java/org/infinispan/util/concurrent/CompletableFutures.java

```diff
@@ -36,6 +40,23 @@
       sink.setFuture(compoundSource);
    }
 
+   public static <T> CompletableFuture<T> connect(NotifyingFuture<T> source) {
+      final CompletableFuture<T> result = new CompletableFuture<>();
+      source.attachListener(f -> {
+         try {
+            result.complete(f.get());
+         } catch (Exception e) {
+            result.cancel(false);
+         }
+      });
+      return result;
+   }
+
+   public static <T> CompletableFuture<List<T>> combine(List<CompletableFuture<T>> futures) {
```

Comment:
@galderz Haskell kind of goes over my head, but isn't `sequence` supposed to be a generic function that can handle any `Monad`?